### DISCUSSION TITLE
migrazione da SD card a SSD

### DISCUSSION THREAD

**Discussion by lorenzopiazzi** [indomus/forum#106] [Hardware & Firmware]
Buongiorno a tutti,
ho cercato di migrare a SSD seguendo la guida di cui a questo link:
https://indomus.it/progetti/avviare-raspberry-pi-tramite-disco-esterno-ssd-o-altro/
Sono sicuro di aver aggiornato il bootloader del mio Pi4 (verificato tramite il comando vcgencmd bootloader_version ), ho usato win32 disk imager per il backup dell'SD e per il suo ripristino sull'SSD e ho inserito in una sd fat32 il bootcode.bin indicato nella guida.
Il problema è che quando poi avvio il raspberry vedo che gira il bootcode ma poi vedo schermo nero e il disco ssd risulta fermo (non vedo lampeggiare il suo led). Senza SD contenente il bootcode il monitor rimane sempre nero.
Ho provato a flashare Raspbian su una chiavetta usb, connetterla al raspberry contenente il bootcode.bin sull'SD formattata fat32 e ha funzionato tutto perfettamente.
Dunque temo che il problema riguardi il restore sull'SSD. Sapete darmi qualche dritta? Da un tool tipo GParted vedo che la struttura sul'SSD è come quella dell'SD originale, ma la partizione che risulta "non allocata" dove stanno i dati di home assistant sia sulla SD che sull'SSD, è di 111GB invece dei 29.7 presenti sull'SD originale... Dovrei partizionare il disco diversamente prima di fare la restore?
Grazie in anticipo!
Lorenzo

P.S. Dal momento che uso Hassio, sapete indicarmi se ripristinare l'.img tramite win32 disk imager è equivalente a flashare l'img "pulita" di Home Assistant sull'SSD e ripristinare poi l'ultima fullsnapshot?

**Comment by NdR91** [comment]
Ciao! Parti dal presupposto che non tutti i chipset dei Box per SSD e/o adattatori sono compatibili con Raspberry pi4.
Potresti provare a bootare con SSD collegato ad una USB 2.0? 

Nel caso in cui funzioni, c'è un sistema abbastanza semplice per fixare il problema ed usarlo sulle USB 3.0...

**Reply by lorenzopiazzi** [reply]
Ciao, grazie mille per la risposta e i suggerimenti!
Purtroppo anche con le USB 2.0 si comporta nella stessa maniera.
Sto provando a creare un'altra immagine dell'SD originale e ripristinarla sull'SSD, tenterò ancora sia con la usb 3.0 che con la usb 2.0, ma nel caso andasse male comprerò un'altra box! Sai suggerirmi un modello che sicuramente funziona o quale criterio seguire nella scelta all'acquisto?
Grazie ancora, non sapevo il problema potesse essere quello.
Lorenzo

**Reply by NdR91** [reply]
Guarda, sinceramente non mi sento di consigliarti un cambio box.
Anzitutto perchè non sai mai davvero che chipset troverai, ma soprattutto perchè il fix esiste e funziona.

Se booti da MicroSD ed inserisci l'ssd esterno, riesci a rilevarlo/montarlo?

**Reply by sky7176** [reply]
Ciao,
io l'unico modo per usare un SSD esterno con PI4 è stato quello di utilizzare la versione a 64bit di Hass.io.

Come case ho utilizzato Argon One M2.

Giuseppe

**Reply by lorenzopiazzi** [reply]
grazie ragazzi per le dritte.
Allora ho provato a seguire il consiglio di @NdR91 e quando collego ad Hassio il disco ssd lui in effetti di qualcosa si accorge (vedi immagine allegata).
Ho provato a flashare l'img di Hassio anche su una chiavetta e si comporta proprio come l'ssd quindi non va al boot mentre viene riconosciuta allo stesso modo se la collego ad hassio già running da SD.
In effetti @sky7176  io ho la versione 32bit.
Deduco che potrebbe essere altrove il problema, non nella box... corretto? 
Sapete suggerirmi la prossima mossa, perché le ho tentate un po' tutte XD 
![Cattura](https://user-images.githubusercontent.com/76661289/113757338-c7aad800-9712-11eb-887b-83dc779bc3ba.PNG)

**Reply by mattapo** [reply]
> Nel caso in cui funzioni, c'è un sistema abbastanza semplice per fixare il problema ed usarlo sulle USB 3.0...

Ciao, io mi trovo nella situazione da te descritta: l'ssd si avvia solo dalle USB2.0 . Mi sapresti indirizzare al fix?

**Reply by NdR91** [reply]
Purtroppo devo ammettere che rileggendo ho notato una cosa che inizialmente mi era sfuggita: Siete su HomeAssistant OS.

Il Fix di cui parlo io, funziona solo su Raspbian (o almeno credo). In ogni caso condivido qui quello che bisogna fare;

Collegare l'SSD nella porta USB 2.0, bootare e cercarlo col comando ```lsusb```. L'output sarebbe qualcosa di simile:

```
pi@raspberrypi:~ $ lsusb
 Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
 Bus 002 Device 002: ID 174c:55aa ASMedia Technology Inc. Name: ASM1051E SATA 6Gb/s bridge, ASM1053E SATA 6Gb/s bridge, ASM1153 SATA 3Gb/s bridge, ASM1153E SATA 6Gb/s bridge
 Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
 Bus 001 Device 002: ID 2109:3431 VIA Labs, Inc. Hub
 Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```
Nell'esempio, quello che ci interessa, è il BUS 002. Nello specifico ci serve il suo ID: **174c:55aa**
A questo punto bisogna abilitare il "Quirk", andando a modificare il file cmdline:

```sudo nano /boot/cmdline.txt```

Aggiungere la seguente riga **all'inizio** del file:

```usb-storage.quirks=XXXX:XXXX:u``` dove al posto delle X bisogna mettere l'ID precedentemente trovato con ```lsusb```.

Il risultato deve essere una cosa simile:

```usb-storage.quirks=174c:55aa:u console=serial0,115200 console=tty1 root=PARTUUID=d34db33f-02 rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait```

Verifica:

```sudo dmesg | grep usb
 [    1.332924] usb 2-1: New USB device found, idVendor=174c, idProduct=55aa, bcdDevice= 1.00
 [    1.332957] usb 2-1: New USB device strings: Mfr=2, Product=3, SerialNumber=1
 [    1.332983] usb 2-1: Product: ASM105x
 [    1.333006] usb 2-1: Manufacturer: ASMT
 [    1.333028] usb 2-1: SerialNumber: 123456789B79F
 [    1.335967] usb 2-1: UAS is blacklisted for this device, using usb-storage instead
 [    1.336071] usb 2-1: UAS is blacklisted for this device, using usb-storage instead
 [    1.336103] usb-storage 2-1:1.0: USB Mass Storage device detected
 [    1.336479] usb-storage 2-1:1.0: Quirks match for vid 174c pid 55aa: c00000 <----- QUESTO DEVE COMPARIRE
 [    1.336611] scsi host0: usb-storage 2-1:1.0
```



@KC-inDomus magari apro un Thread dedicato con tutte queste istruzioni, che ne pensi?

**Reply by mattapo** [reply]
Scusa, ho omesso il fatto che io uso core con raspbian. Grazie per le info, provo subito!

**Reply by mattapo** [reply]
Perfetto, funziona alla grande su Raspbian! Così facendo utilizza tutta la banda, e quindi la velocità, dell'usb 3.0?

**Reply by NdR91** [reply]
NI. Di fatto castri un pelino la banda offerta dall'SSD, che rimane comunque decisamente superiore sia alla MicroSD, sia che all'eventuale connessione ad USB 2.0.

Se ricordo bene (è passato un pò), avevo testato le velocità in scrittura e lettura e si aggiravano sui 300/400 MB/s (anzichè i 600 previsti da un SSD SATA).

Per come la vedo io è comunque perfetto per un Raspberry, che dubito trarrebbe reale beneficio tra i 400 ed i 600 MB/s. Il Reale salto lo fai dai 30/40 di una MicroSD ai 300 e passa dell'SSD.

Se comunque sei curioso, ti consiglio questo script:
[Storage.sh](https://raw.githubusercontent.com/TheRemote/PiBenchmarks/master/Storage.sh)

Per usarlo ti basta un curl:
```sudo curl https://raw.githubusercontent.com/TheRemote/PiBenchmarks/master/Storage.sh | sudo bash```

**Comment by sky7176** [comment]
Io proverei con la versione a 64bit

Il mar 6 apr 2021, 20:01 lorenzopiazzi ***@***.***> ha
scritto:

**Reply by lorenzopiazzi** [reply]
Proverò con la 64bit. Quindi suggerite di flashare hassio 64bit sull'SSD e se viene letto procedere a ripristinare l'ultima full-snapshot? 
Sapete dirmi se il ripristino della full-snapshot equivalga esattamente il caricamento dell'.img eseguito con win32 disk imager?
Grazie ancora per il supporto,
Lorenzo

**Comment by essegi65** [comment]
anche io ho usato un RPI4 a 64 bit. Eseguito l'aggiornamento con del bootloader con raspberry pi imager (con aggiornamento avvenuto vedi lo schermo del video collegato VERDE), usandolo anche per la scrittura dell'immagine sul nuovo SSD
